options to configure static relays for autorelay

vyzo opened the pull request:
So that we can avoid hitting the DHT; also changes the desired relays parameter to 1 (from 3).
See #694

albrow commented:
Does the `StaticRelays` option mean that those are the *only* relays that can be used? Ideally, for our network we would like to use static relays to reduce the time to first connection but we still want to discover other relays over time. Otherwise our dedicated relay hosts would be overworked.

vyzo commented:
With this patch yes, when specified only those will be used.
Perhaps we should have a fallback to the DHT, but the issue is that we are having excessive DHT traffic because of autorelay.

vyzo commented:
ping @Stebalien

vyzo commented:
Moving forward we also want to use relays discovered by passive discovery (when the option `circuit.OptDiscovery` is specified).

raulk commented on options.go:
```diff
@@ -245,6 +246,39 @@ func EnableAutoRelay() Option {
+		for _, addr := range []string{
+			"/ip4/147.75.80.110/tcp/4001/p2p/QmbFgm5zan8P6eWWmeyfncR5feYEMPbht5b1FW1C37aQ7y",
+			"/ip4/147.75.195.153/tcp/4001/p2p/QmW9m57aiBDHAkKj9nmFSEn7ZqrcF1fZS4bipsTCHburei",
+			"/ip4/147.75.70.221/tcp/4001/p2p/Qme8g49gm3q4Acp7xWBKg3nAa9fxZ1YmyDJdyGgoG6LsXh",
```
Can we have these multiaddrs in a global var? You can initialize them with an IIFE, or in an init function.

vyzo replied:
sure.

raulk commented on p2p/host/relay/autorelay.go:
```diff
@@ -48,12 +50,13 @@ type AutoRelay struct {
 }
 
-func NewAutoRelay(ctx context.Context, bhost *basic.BasicHost, discover discovery.Discoverer, router routing.PeerRouting) *AutoRelay {
+func NewAutoRelay(ctx context.Context, bhost *basic.BasicHost, discover discovery.Discoverer, router routing.PeerRouting, static []peer.AddrInfo) *AutoRelay {
```
This starts screaming "functional options".

vyzo replied:
*shrug*

vyzo replied:
it's not really user visible, so it doesn't offer much in terms of ergonomics.

raulk commented on p2p/host/relay/autorelay.go:
```diff
@@ -245,6 +248,10 @@ func (ar *AutoRelay) connect(ctx context.Context, pi peer.AddrInfo) bool {
 }
 
 func (ar *AutoRelay) discoverRelays(ctx context.Context) ([]peer.AddrInfo, error) {
+	if len(ar.static) > 0 {
```
I don't think setting static relays should be exclusive from dynamic relay discovery. You probably want static relays for a speedier bootup, but as you dynamically discover relays, you should preferentially switch to those.

raulk replied:
I see that @albrow expected the same thing: https://github.com/libp2p/go-libp2p/pull/705#issuecomment-522708344

vyzo replied:
This defeats the purpose; the main impetus for the static relay configuration is to avoid hitting the DHT _altogether_.
We might want to leverage passive discovery with `CanHop` in the relay module, but I don't think it's a good idea to have both static relays _and_ hit the dht.

vyzo commented:
Updated to use a global variable in the autorelay module for the default relays; also renamed the `DefaultRelays` to `DefaultStaticRelays` and made `StaticRelays` append instead of overwriting.

Stebalien commented:
Agreed. Static relays are really just an emergency measure until we can implement a better system.

RubenKelevra commented:
> Agreed. Static relays are really just an emergency measure until we can implement a better system.

How about flipping the process, having a node subscribe to pubsub when they need a relay.

All nodes which offer the relay service will send a message on this pubsub on how many connections they can accept.

The number is multiplied with a number x like 0.01 in the beginning.

On connection to the relay we send a 'request' for relay services. When the relay is already full, it will send a deny message and close the connection.

On each failure to connect to a relay, the number x is cut down in half, to simply estimate the 'load' on all the relays.

After a random period between 30 and 600 minutes the number x is multiplied by 2 again.

Stebalien commented:
We should advertise relays through pubsub (and maybe even just have relays advertise other relays). However, there's still the issue of reputation.

RubenKelevra commented:
@Stebalien I think that's an issue for another time. I think it's best to just replace the DHT method with something that's responding dynamically to the network size, by lowering the connection attempts.

The optimal solution for fixing this would be to add peers in a Web Of Trust fashion and allow 'friends' to use the relay function.

We could also hold those connections all the time and bootstrap with the friend connections first, instead of the static IPs and domains we currently use.

Stebalien commented:
I'm not sure if that's an issue we _can_ punt on. It takes a bit of effort to attack the DHT. It would take very little effort to add enough bad relays to affect the network.

We don't really need a full WoT system, even just something that verifies that a relay is doing the right thing would be enough. For example, some way to ask other peers to dial back through the relay and verify that it's working.

RubenKelevra commented:
@Stebalien if we change the amount of desired relays back to 3, we could check each of those relays by requesting some data:

One relay is asked to provide some random stuff, and we request it thru both other relays.

We cycle this thru all 3 relays to confirm all relays are working fine.

---

Another possibility is to request everything round robin thru the 3 relays, and if we identify a relay which is sending data which doesn't match the checksums for example, we drop that connwc and create a new one.

aschmahmann commented on p2p/host/relay/autorelay.go:
```diff
@@ -48,12 +57,13 @@ type AutoRelay struct {
 		discover:   discover,
 		router:     router,
 		addrsF:     bhost.AddrsFactory,
+		static:     static,
```
@vyzo I know I'm late to the party here, but this felt a bit small for an issue.

Why did we bother changing the constructor and using a `[]Addr.Info` here instead of just creating a `Discoverer` implementation that returns a the static list?

Sure it can be more efficient to use the slice, but now we have this thing we're going to be stuck removing.
